// lotin-kirill/dist-src/utils.js
import anchorme from 'anchorme';
export function isLowerCase(char) {
  return char.toLowerCase() === char;
}
export function normalizeCase(char, assimilateTo) {
  return isLowerCase(assimilateTo) ? char.toLowerCase() : char.toUpperCase();
}
export function capitalize(word) {
  return word.charAt(0).toUpperCase() + word.substring(1).toLowerCase();
} // Detect only alphabetical letters, excluding apostrophe(')

export function isLetter(char) {
  if (typeof char === 'string') {
    return /^[a-z\u0400-\u04FF]+$/i.test(char);
  }

  return false;
}
export function isVovel(char) {
  return /[аоэеиуўёюяaeiou]/i.test(char);
}
export function isURL(word) {
  return anchorme.list(word).length > 0;
}
export function escapeRegex(expString) {
  return expString.replace(/[-\/\\^$*+?.()|[\]{}]/g, '\\$&');
}
export function endsWithSoftSign(word) {
  return /ь$/i.test(word);
}
export function isRomanNumber(word) {
  return /^M{0,4}(CM|CD|D?C{0,3})(XC|XL|L?X{0,3})(IX|IV|V?I{0,3})(\.|\))?$/.test(word);
}
